### PR TITLE
rgw_sal_motr: [CORTX-30178] Fix for listing object versions in lexicographical order

### PR DESCRIPTION
**Pull request by priyanka25081999**
Problem description:
* Object versions listed in random order in List Object Versions API

Solution:
* Sort the object versions based on last_modified_time of objects in MotrBucket::list() method.

Signed-off-by: Priyanka Salunke <priyanka.s.salunke@seagate.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1034,6 +1034,19 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
     }
   }
 
+  // Sort the object versions
+  vector<rgw_bucket_dir_entry>::iterator iter;
+
+  for (iter = results.objs.begin(); iter != results.objs.end(); ++iter) {
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:1040]: why are we iterating over `results.objs`??
Just the `sort()` function should do r8?

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -1034,6 +1034,19 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
     }
   }
 
+  // Sort the object versions
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:1037]: // Sort by object versions

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -1034,6 +1034,19 @@ int MotrBucket::list(const DoutPrefixProvider *dpp, ListParams& params, int max,
...
+  vector<rgw_bucket_dir_entry>::iterator iter;
+
+  for (iter = results.objs.begin(); iter != results.objs.end(); ++iter) {
+    std::sort(results.objs.begin(), results.objs.end(), [](const rgw_bucket_dir_entry& res1, const rgw_bucket_dir_entry& res2)
+    {
+      if (res1.key.name == res2.key.name)
+        return res1.meta.mtime > res2.meta.mtime;
+      else
```
Comment (sumedhak27) [inline src/rgw/rgw_sal_motr.cc:1045]: suggestion: 
no need of `else` case, just a return will work.

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc:1040]: yeah, lamda function would iterate through whole list anyways !?

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc:1045]: what's the intention here @priyanka25081999, to sort by key.name in descending order, why do we need further mtime( modtime i guess) check, if key.name are same, order would matter there ?

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (priyanka25081999) [reply src/rgw/rgw_sal_motr.cc]: Yes, I had added iterator to do more debugging.
We can simply add sort() function. 
Done all changes as suggested.

**Comment by priyanka25081999** [comment]
We found another approach to handle this ordering issue and seems to be effective, hence closing this PR.

**Pull request closed without merging**